### PR TITLE
vm: make `types` list append-only

### PR DESCRIPTION
**Pull request by zerbina**
## Summary
A separate hash-table overlay is used for looking up structural types, making
the previous list re-orderings used for speeding up the linear search obsolete.

This is an important step towards replacing `PVmType` with integer IDs

### Future work:
- `genTuple` should also return the already computed structural-type hash
- the equality function used by `getOrIncl` should be made override-able

### CODE REVIEW

File: compiler/vm/vmtypegen.nim [hunk 1]
```diff
@@ -38,6 +38,71 @@ template paddedSize(info: VmType): untyped =
...
+    of akSet:                 hash(t.setLength)
+    of akCallable, akClosure: hash(t.funcTypeId.int)
+    of akDiscriminator:       hash(t.numBits)
+    of akArray:               hash(t.elementCount) !& hash(t.elementType)
+    of akObject:
+      var r = 0
+      for (_, ft) in t.objFields.items:
+        r = r !& hash(ft)
+      r
+    of akInt, akFloat, akString, akPNode:
```
Comment (saem) [inline compiler/vm/vmtypegen.nim:62]: So we wouldn't match a type who's first element is an int with an int parameter in a function?

**Comment by saem** [comment]
Left a comment if a non issue then we're good to go.

Bors d

**Comment by saem** [comment]
Bors d+

File: compiler/vm/vmtypegen.nim (hunk 1 above)
Comment (zerbina) [reply compiler/vm/vmtypegen.nim:62]: No, they wouldn't match. Procedural types use a separate process that turns them into unique IDs first (`getOrCreateFuncType`)

**Pull request merged**